lib/saml_authenticator.rb

```diff
@@ -205,51 +207,69 @@ def auto_create_account(result, uid)
+    groups_attributes = setting(:groups_attribute).split(",")
+    group_match_column = setting(:groups_use_full_name) ? "full_name" : "name"
+
+    groups_from_groups_attributes =
+      groups_attributes
+        .flat_map { |attr| attributes.multi(attr.strip) || [] }
+        .compact
+        .map { |g| g.downcase.split(",") }
+        .flatten
+        .uniq
 
     if setting(:groups_ldap_leafcn).present?
       # Change cn=groupname,cn=groups,dc=example,dc=com to groupname
-      user_group_list = user_group_list.map { |group| group.split("=", 2).last }
+      groups_from_groups_attributes =
+        groups_from_groups_attributes.map { |group| group.split("=", 2).last }
     end
 
     if groups_fullsync
-      user_has_groups = user.groups.where(automatic: false).pluck(:name).map(&:downcase)
-      groups_to_add = user_group_list - user_has_groups
-      groups_to_remove = user_has_groups - user_group_list if user_has_groups.present?
-    else
-      total_group_list = setting(:sync_groups_list).split("|").map(&:downcase)
+      user_has_groups =
+        user.groups.where(automatic: false).pluck(group_match_column).compact.map(&:downcase)
 
+      groups_to_add = groups_from_groups_attributes - user_has_groups
+      groups_to_remove = user_has_groups - groups_from_groups_attributes if user_has_groups.present?
+    else
       groups_to_add = info["groups_to_add"] || attributes.multi("groups_to_add")&.join(",") || ""
       groups_to_add = groups_to_add.downcase.split(",")
-      groups_to_add += user_group_list
+      groups_to_add += groups_from_groups_attributes
 
       groups_to_remove =
         info["groups_to_remove"] || attributes.multi("groups_to_remove")&.join(",") || ""
       groups_to_remove = groups_to_remove.downcase.split(",")
 
-      if total_group_list.present?
-        groups_to_add = total_group_list & groups_to_add
+      groups_from_attributes_before_sync =
+        get_groups_from_attributes(user:, groups_attributes:, previous_attributes:)
+      groups_to_remove += (groups_from_attributes_before_sync - groups_from_groups_attributes)
+
+      groups_allowlist = setting(:sync_groups_list).split("|").map(&:downcase)
+      if groups_allowlist.present?
+        groups_to_add = groups_allowlist & groups_to_add
 
         removable_groups = groups_to_remove.dup
-        groups_to_remove = total_group_list - groups_to_add
+        groups_to_remove = groups_allowlist - groups_to_add
         groups_to_remove &= removable_groups if removable_groups.present?
       end
     end
 
-    return if user_group_list.blank? && groups_to_add.blank? && groups_to_remove.blank?
+    return if groups_to_add.blank? && groups_to_remove.blank?
 
-    Group
-      .where("LOWER(name) IN (?) AND NOT automatic", groups_to_add)
-      .each { |group| group.add user }
+    if groups_to_add.present?
+      Group
+        .where("LOWER(#{group_match_column}) IN (?) AND NOT automatic", groups_to_add)
+        .each { |group| group.add user }
```

Comment:
This is not related to your change but I think it is useful to point out that there is potential for abuse here since we have no limits on how groups a user can be added to a group each time.